protocol/mqtt_paho/v2/context.go

```diff
@@ -0,0 +1,43 @@
+package mqtt_paho
+
+import "context"
+
+// ProtocolContext allows a Receiver to understand the context of a request.
+type ProtocolContext struct {
+	Topic    string
+	QoS      byte
+	Retained bool
+}
+
+func NewProtocolContext(topic string, qos byte, retained bool) ProtocolContext {
+	return ProtocolContext{
+		Topic:    topic,
+		QoS:      qos,
+		Retained: retained,
+	}
+}
+
+// Opaque key type used to store ProtocolContext
+type protocolContextKeyType struct{}
+
+var protocolContextKey = protocolContextKeyType{}
+
+// WithProtocolContext return a context with the given ProtocolContext into the provided context object.
+func WithProtocolContext(ctx context.Context, pctx ProtocolContext) context.Context {
+	return context.WithValue(ctx, protocolContextKey, pctx)
+}
+
+// ProtocolContextFrom pulls a ProtocolContext out of a context. Always
+// returns a non-nil object.
+func ProtocolContextFrom(ctx context.Context) ProtocolContext {
```

Comment:
Just wondering if it would be better to return nil if there isn't one and then we can avoid the deepEqual below and just check against nil?

Reply:
+1 make pointer type and return &val

Reply:
@duglin @embano1 Thanks for the reminder, it's better to use a pointer!